state-chain/pallets/cf-broadcast/src/lib.rs

```diff
@@ -793,7 +787,7 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			// If the signature verification fails, we want
 			// to retry from the threshold signing stage.
 			else {
-				Self::clean_up_broadcast_storage(broadcast_id);
+				Self::clean_up_broadcast_storage(broadcast_attempt);
```

Comment:
Btw. not sure this *introduces* a bug? If anything the bug was already present: we were already cleaning up all the other state including the callback.

If this edge case is a concern then we should avoid cleaning up anything used in broadcast_succeeded.

Another way around this might be to use external chain blocks for expiry rather than SC blocks.

I raised PRO-989 to mitigate this.

Reply:
> This case should not happen after the broadcast barriers functionality.

Think it's still an issue even after the barrier, because we're talking about a tx that began it's broadcast *before* the rotation was initiated. 

> Btw. not sure this introduces a bug?

True, actually was a bug before.

> we should avoid cleaning up anything used in broadcast_succeeded

👍  

> Another way around this might be to use external chain blocks for expiry rather than SC blocks.

Even if it used on chain blocks for expiry this is still an issue? Not sure how that helps things.

Reply:
> Think it's still an issue even after the barrier, because we're talking about a tx that began it's broadcast _before_ the rotation was initiated.
 
Yes but the first barrier (in eth's case) will make sure that all the txs preceeding the rotation tx have successfully witnessed BEFORE even initiating the rotation tx. For polkadot's case, we never have to resign a tx as the sig is always valid around a rotation and so we dont go into the else arm where the cleanup is done. 

In polkadot, we only go into the else arm (the tx has to be resigned) in a rarae case of when polkadot runtime version has been updated. But in that case, the tx will always have to be resigned and an earlier version cannot succed on polkadot chain so this case should not happen in that scenario. An only very very unlikely case is still possible where these mempool issues you are talking about (which anyways are rare to affect only one tx and not others) happen at the exact time of a polkadot runtime version update (which itself is a rare event) so its propably ok.

For Bitcoin, it would never go into the else arm since the broadcast is always valid and never has to be resigned.

Reply:
Also, the edge case mentioned above results in other bigger problems for example in ```transaction_succeeded``` we have 

```rust
let to_refund = AwaitingBroadcast::<T, I>::get(BroadcastAttemptId {
	broadcast_id,
	attempt_count: BroadcastAttemptCount::<T, I>::get(broadcast_id),
	})
	.ok_or(Error::<T, I>::InvalidBroadcastAttemptId)?
```

which assumes that the succeeded attempt was the last attempt. This would error in the edge case mentioned.

Reply:
> Also, the edge case mentioned above results in other bigger problems for example in `transaction_succeeded` we have
> 
> ```rust
> let to_refund = AwaitingBroadcast::<T, I>::get(BroadcastAttemptId {
> 	broadcast_id,
> 	attempt_count: BroadcastAttemptCount::<T, I>::get(broadcast_id),
> 	})
> 	.ok_or(Error::<T, I>::InvalidBroadcastAttemptId)?
> ```
> 
> which assumes that the succeeded attempt was the last attempt. This would error in the edge case mentioned.

As discussed, this should be fine since the latest attempt's Awaiting broadcast should always exist and we only need transaction_payload from this storage item which does not depend on the attempt.